spec/system/consumer/split_checkout_spec.rb

```diff
@@ -107,23 +107,95 @@
         end
 
         it "redirects the user to the Payment Method step" do
-          fill_in 'Any comments or special instructions?', with: "SpEcIaL NoTeS"
-          click_button "Next - Payment method"
-          expect(page).to have_current_path("/checkout/payment")
+          fill_notes("SpEcIaL NoTeS")
+          proceed_to_payment
         end
       end
 
-      describe "selecting a delivery shipping method and submiting the form" do
+      describe "selecting a delivery method" do
         before do
           choose shipping_with_fee.name
-          uncheck "ship_address_same_as_billing"
         end
 
-        it "redirects the user to the Payment Method step" do
-          fill_out_shipping_address
-          fill_in 'Any comments or special instructions?', with: "SpEcIaL NoTeS"
-          click_button "Next - Payment method"
-          expect(page).to have_current_path("/checkout/payment")
+        context "with same shipping and billing address" do
+          before do
+            check "ship_address_same_as_billing"
+          end
+          it "does not display the shipping address form" do
+            within(:xpath, './/div[@class="checkout-substep"][3]') do
+              expect(page).not_to have_field "order_ship_address_attributes_address1"
+            end
+          end
+
+          it "redirects the user to the Payment Method step, when submiting the form" do
+            proceed_to_payment
+            # asserts whether shipping and billing addresses are the same
+            ship_add_id = Spree::Order.first.ship_address_id
+            bill_add_id = Spree::Order.first.bill_address_id
+            expect(Spree::Address.where(id: bill_add_id).pluck(:address1) ==
+              Spree::Address.where(id: ship_add_id).pluck(:address1)).to be true
+          end
+        end
+
+        context "with different shipping and billing address" do
+          before do
+            uncheck "ship_address_same_as_billing"
+          end
+          it "displays the shipping address form and the option to save it as default" do
+            within(:xpath, './/div[@class="checkout-substep"][3]') do
+              expect(page).to have_field "order_ship_address_attributes_address1"
+            end
+          end
+
+          it "displays error messages when submitting incomplete billing address" do
+            click_button "Next - Payment method"
+            expect(page).to have_content "Saving failed, please update the highlighted fields."
+            within(:xpath, './/div[@class="checkout-substep"][3]') do
```

Comment:
Hmmm... I'm not a big fan of all these uses of `within (:xpath, './/div[@class="checkout-substep"][3]') do` everywhere with different numbers. It feels like we're creating really tight coupling here between the tests and the minute details of the layout of the page, which means in the future any time we make small changes in the views we'll potentially have to rewrite all these tests.

I just had a quick look, and I think almost all of these tests still pass if we remove these uses of `within(:xpath ...)` and the remaining one or two tests work fine with a very slight modification.

I'd prefer if we went in that direction here...

Reply:
I'm not a fan either :-) I [share the concern (point 2)](https://github.com/openfoodfoundation/openfoodnetwork/pull/8677#issuecomment-1006474515), I felt this approach was needed to make a distinction between bill and ship address fields, as identical `css` labels.

But thinking it again I guess we don't need the `within` blocks as each field seems to have its own specific `id`s, like `order_ship_address_attributes_address1` vs `order_bill_address_attributes_address1`.